packages/runtime-core/__tests__/apiWatch.spec.ts

```diff
@@ -44,6 +45,28 @@ describe('api: watch', () => {
     expect(dummy).toBe(1)
   })
 
+  // https://github.com/vuejs/vue-next/issues/3322
+  it('effect onInvalidate succeeds in ssr context', done => {
+    // XXX do this without double setup and generally less hackily!
+    let instance: ComponentInternalInstance | null
+    const noop = () => {}
+    const Comp = defineComponent({
+      setup() {
+        if (!instance) {
+          instance = getCurrentInstance()
+        } else {
+          watchEffect(onInvalidate => {
+            expect(() => onInvalidate(noop)).not.toThrow(ReferenceError)
+            done()
```

Comment:
This will get called by the first watchEffect boostrap, which is not the SSR one, so an error in the SSR one would not be part of the test anymore, I think?

Reply:
I'm assuming (and running the tests seems to validate) that setup gets called twice. The first time, instance is not set, so we just set `instance`. The second setup call with SSR triggers the `watchEffect` branch.

I could rewrite with promises but I think it would be a bit more verbose... is that preferred?